### PR TITLE
Implementation of optimistic lock control to mongodb-panache, both reactive and non reactive

### LINKED ISSUE #15124: Optimistic lock control on Panache MongoDB
Hi guys, hope you are doing well.

I was thinking about something I've commented in another ticket some time ago, but in that ticket we are talking about MongoDB migrations. 
What about we include in MongoDB Panache an optimistic lock feature, just like we have in JPA. Spring and other frameworks already have some similar.

I think it is not so complicated to implement, basically we need an annotation `@Version` or some similar to identify the field used as "version", and some adjusts in `MongoOperations` the methods `update`, and `persistOrUpdate`.

Today when we call `update`, in the `MongoOperations.update` we build the query:

```
BsonValue id = document.get(ID);
BsonDocument query = new BsonDocument().append(ID, id);
```

We can change for some like:

```
BsonValue id = document.get(ID);
BsonValue version = document.get(VERSION);
return new BsonDocument().append(ID, id).append(VERSION, version);
```

In addition to that we add an `OptimisticLockException` to inform users trying to update outdated documents.

Do you think it can work?  

Best.

### PR DESCRIPTION
**Pull request by diogocarleto**
This PR is intent to bring optimist lock control to mongodb-panache, similar what we have in JPA/Hibernate. It is both supported for reactive and non reactive.

To use that we just need to use the annotation `@Version` in a entity field, such as:

```
public class Person {
  public String name

  @Version
  public Long version;
}
```

The optimistic lock control is supported for the methods `persist/persistOrUpdate/update` for one single entity or many entities, we can check that in the tests in `io.quarkus.it.mongodb.panache.OptimisticLockControlRepositoryIntegrationTests` and `io.quarkus.it.mongodb.panache.OptimisticLockControlEntityIntegrationTests`.

The changes mainly resides in `io.quarkus.mongodb.panache.common.runtime.MongoOperations`.

Like in jpa/hibernate, every persist/update will increment the `version` value.
If an update fail a `io.quarkus.mongodb.panache.common.exception.OptimisticLockException` is raised, also similar to jpa/hibernate. 

Let me know what do you think and improvements we should do here.

Fix #15124

### CODE REVIEW

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    public static Long extractVersionValue(Field versionField, Object entity) {
+        try {
+            if (versionField == null) {
+                return null;
+            }
+
+            boolean canAccess = versionField.canAccess(entity);
+            versionField.setAccessible(true);
+            Long versionValue = (Long) versionField.get(entity);
+            versionField.setAccessible(canAccess);
```
Comment (evanchooly) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:60]: This setting of 'accessible' back and forth is unnecessary.  Calling `setAccessible()` only affects the `Field` instance (in this case) that it was called on. Code else where that extracts a `Field` reflectively gets a different instance so there is no cross-contamination risk here.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            return versionValue;
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Error on set the version value");
```
Comment (evanchooly) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:63]: Should probably include the error message from the IAE.

**Review by evanchooly — COMMENTED** [review]
In general, I have objections to this feature for various reasons I've stated before but I have some comments on what's here at least. One thing I noticed is there are no kotlin tests. Reflection is slightly different in kotlin itself but perhaps what's here will work just fine with kotlin generated bytecode.  It probably will. But there should be tests to determine that as well.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    private static Field extractVersionField(Object entity) {
+        Field[] fields = entity.getClass().getFields();
+        Field[] declaredFields = entity.getClass().getDeclaredFields();
+        return Stream.of(fields, declaredFields)
+                .flatMap(Stream::of)
+                .filter(field -> field.isAnnotationPresent(Version.class))
+                .findFirst().orElse(null);
+    }
+
+    public static Long extractVersionValue(Field versionField, Object entity) {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:51]: doesn't make sense, private is enough.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 2 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:63]: sorry for the question, what is IAE? @evanchooly is there any slack channel or other place I can ping you for future discussions of the features, rather than we discuss here in the PR? If not, no problems at all.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 4]
```diff
@@ -33,19 +34,23 @@
...
+import io.quarkus.mongodb.panache.common.exception.OptimisticLockException;
+import io.quarkus.mongodb.panache.common.runtime.util.VersionHandler;
 import io.quarkus.mongodb.panache.common.transaction.MongoTransactionException;
 import io.quarkus.panache.common.Parameters;
 import io.quarkus.panache.common.Sort;
 
 @SuppressWarnings({ "rawtypes", "unchecked" })
 public abstract class MongoOperations<QueryType, UpdateType> {
     public static final String ID = "_id";
+    private static final String VERSION = "version";
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: you are right, I've changed it, and now any variable name annotated with @Version will works.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 5]
```diff
@@ -175,15 +180,20 @@ public void persistOrUpdate(Stream<?> entities) {
...
+        DeleteResult deleteResult;
         if (session == null) {
-            collection.deleteOne(query);
+            deleteResult = collection.deleteOne(query);
         } else {
-            collection.deleteOne(session, query);
+            deleteResult = collection.deleteOne(session, query);
+        }
+
+        if (versionHandler.containsVersionAnnotation() && deleteResult.getDeletedCount() == 0) {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: changed.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 6]
```diff
@@ -206,6 +216,12 @@ public MongoDatabase mongoDatabase(Class<?> entityClass) {
 
     private void persist(MongoCollection collection, Object entity) {
         ClientSession session = getSession(entity);
+
+        VersionHandler versionUtil = VersionHandler.of(entity);
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: done.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 7]
```diff
@@ -263,16 +291,30 @@ private void persistOrUpdate(MongoCollection collection, Object entity) {
...
         } else {
-            //insert with user provided ID or update
-            BsonDocument query = new BsonDocument().append(ID, id);
-            if (session == null) {
-                collection.replaceOne(query, entity, new ReplaceOptions().upsert(true));
-            } else {
-                collection.replaceOne(session, query, entity, new ReplaceOptions().upsert(true));
+            VersionHandler versionUtil = VersionHandler.of(entity);
+            if (!versionUtil.containsVersionAnnotation()) {
+                BsonDocument query = buildUpdateQueryWithoutVersion(entity);
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: done.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            versionField.setAccessible(true);
+            Long versionValue = (Long) versionField.get(entity);
+            versionField.setAccessible(canAccess);
+            return versionValue;
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Error on set the version value");
+        }
+    }
+
+    public void adjustVersionValue() {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java]: adjusted

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 1 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java]: changed.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/Version.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.FIELD })
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/Version.java:11]: I understand your point of view, but please, see my point of view. 

I've started with Panache almost 3 years ago, and even at that time, the guides were excellent, Panache was really easy to use, panache queries make easier for the developers to write queries similar to sql ansi, there were no out of box configuration to have all working, so for sure, others started using it just like me, and in some moment realized that they needed optimistic lock control (or even audit, expurge), for sure others have similar implementation like I did in my job.

It is really complicated if we say to the community that have projects in production with +100 entities looking for having optimistic lock control, "hey panache is a tiny layer and we are not adding this here, please use the tool/framework/extension xyz". Do you agree it doesn't make sense? If we decided to use any approach similar in the future, we must give the community the way to easy migrate without pain.

I'm really grateful for all the features that has been added day after day to Quarkus, it is amazing, it makes Quarkus great, the fast pace of the evolution, the community embracing the project, according with the needs of the community.

I'm here to contribute and help making Quarkus even bigger.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: adjusted

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 10]
```diff
@@ -309,6 +360,47 @@ private void persistOrUpdate(List<Object> entities) {
...
+        BsonDocument document = getBsonDocument(collection, entity);
+
+        //then we get its id field and create a new Document with only this one that will be our replace query
+        BsonValue id = document.get(ID);
+
+        VersionHandler versionUtil = VersionHandler.of(entity);
+        if (versionUtil.containsVersionAnnotation()) {
+            BsonValue version = document.get(VERSION);
+            if (version == null) {
+                version = new BsonNull();
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: because If I send an entity to update without version value, we will wrongly update an entity that has the version 0. Because we inferred 0 here.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        BsonDocument query = buildUpdateQuery(versionFieldName, document);
+
+        return new VersionHandler(entity, query, versionField, true, versionValue != null, versionFieldName);
+    }
+
+    public boolean containsVersionAnnotationAndValue() {
+        return containsVersionAnnotation && containsVersionValue;
+    }
+
+    private static Optional<Field> extractVersionField(Object entity) {
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:70]: It would be better to use Jandex at build time to compute a map of version field by class, this would avoid using reflection at runtime which we want to avoid.
I also don't like the fact that if multiple fields exist in the class hierarchy you just take the first one, you should either use the closest one on the hierarchy or generate an error.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 2 above)
Comment (loicmathieu) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:89]: +1, IAE = IllegalAccessException

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+    /**
+     * Utility method to check if the during an update there are documents not updated.
+     * Throws OptimisticLockException when found documents not updated.
+     */
+    public void checkHasNotAffectedResults(Long quantityResultAffected) {
+        if (containsVersionAnnotation && quantityResultAffected == 0) {
+            throwOptimisticLockException(entity);
+        }
+    }
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:115]: Duplicated with the same version inside the version handler class.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java [hunk 13]
```diff
@@ -208,9 +211,12 @@ public Uni<Void> persistOrUpdate(Stream<?> entities) {
...
         BsonDocument document = getBsonDocument(collection, entity);
-        BsonValue id = document.get(ID);
-        BsonDocument query = new BsonDocument().append(ID, id);
-        return collection.deleteOne(query).onItem().ignore().andContinueWithNull();
+        VersionHandler versionHandler = VersionHandler.handle(entity, document);
+
+        return collection.deleteOne(versionHandler.query).onItem()
+                .invoke(deleteResult -> versionHandler
+                        .checkHasNotAffectedResults(((DeleteResult) deleteResult).getDeletedCount()))
+                .onItem().ignore().andContinueWithNull();
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:219]: Is it really related to optimistic lock controll to verify that a delete operation really delete something? I'm not very compfortable and more important: if I want to delete something and someone else already deleted it why throw ?

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java [hunk 14]
```diff
@@ -235,25 +241,40 @@ public Uni<Void> nullUni() {
     }
 
     private Uni<Void> persist(ReactiveMongoCollection collection, Object entity) {
+        VersionHandler.handle(entity, getBsonDocument(collection, entity));
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:244]: At first, I thought it was a no-op, then I see that VersionHandle.handle creates a version handle wich modify the entity inside it's adjust method called by the constructor.
I find the code path very hard to follow and by looking at the modification inside the MongoOperation class we don't very inderstand what's going on.
Think that many people should open this class without knowledge of how the version mechanism works, this should be more straigthforward to understand.

**Review by loicmathieu — COMMENTED** [review]
Hi,
Thanks for working on this.
I made a first round of review but as there is a lot of changes I didn't had time to look in details and I skip all the tests.

As @evanchooly I'm not a big fan of such big addition to MongoDB with Panache but I understand the rational here.

Also, all the version checks are done after the database was updated, so it can only works when using transactions right ? If so, it must be checked and an error must be thrown if misused.

**Review by evanchooly — COMMENTED** [review]
I understand the desire to have something like this but I still contend that Panache is not the place for this.  I have a number of issues/disagreements with the implementation as it stands now most of which I've left comments about.  I won't personally "block" this (I don't think I even *can*, really) but I don't see myself approving the PR either.  I understand a lot of work has gone in to it but this is really the wrong place for this feature.  If you really need optimistic locking please consider something like [quarkus-morphia](https://github.com/quarkiverse/quarkus-morphia) which has a fully functional, battle tested implementation.

**Review by evanchooly — CHANGES_REQUESTED** [review]
changes in prior review as well as general objections

File: docs/src/main/asciidoc/mongodb-panache-kotlin.adoc [hunk 15]
```diff
@@ -198,3 +198,30 @@ implementation("io.quarkus:quarkus-mongodb-panache-kotlin") <1>
 ----
 <1>  Note the addition of `-kotlin` on the end.  Generally you'll only need this version but if your project will be using
 both Java and Kotlin code, you can safely include both artifacts.
+
+== Optimistic Lock Control
+
+Like we have in JPA/Hibernate, optimistic lock control is now supported, for this you just need to use the `@Version`, and panache does the rest.
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply docs/src/main/asciidoc/mongodb-panache-kotlin.adoc:204]: I see, I'll improve this.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 16]
```diff
@@ -309,6 +360,47 @@ private void persistOrUpdate(List<Object> entities) {
...
+    private BsonDocument buildUpdateQuery(Object entity) {
+        MongoCollection collection = mongoCollection(entity.getClass());
+        //we transform the entity as a document first
+        BsonDocument document = getBsonDocument(collection, entity);
+
+        //then we get its id field and create a new Document with only this one that will be our replace query
+        BsonValue id = document.get(ID);
+
+        VersionHandler versionUtil = VersionHandler.of(entity);
+        if (versionUtil.containsVersionAnnotation()) {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java]: done

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java [hunk 17]
```diff
@@ -295,10 +337,19 @@ private void persistOrUpdate(List<Object> entities) {
                 //insert with autogenerated ID
                 bulk.add(new InsertOneModel(entity));
             } else {
-                //insert with user provided ID or update
-                BsonDocument query = new BsonDocument().append(ID, id);
-                bulk.add(new ReplaceOneModel(query, entity,
-                        new ReplaceOptions().upsert(true)));
+                //insert/update with user provided ID or update
+                VersionHandler versionHandler = VersionHandler.of(entity);
+                if (versionHandler.containsVersionAnnotationAndValue()) {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/MongoOperations.java:328]: done

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 2 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:89]: I see, done

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/Version.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
+package io.quarkus.mongodb.panache.common;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Documented
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/Version.java:9]: sure, done

File: docs/src/main/asciidoc/mongodb-panache-kotlin.adoc (hunk 15 above)
[the remark this answers is not in the document]
Comment (diogocarleto) [reply docs/src/main/asciidoc/mongodb-panache-kotlin.adoc:204]: done

File: docs/src/main/asciidoc/mongodb-panache.adoc [hunk 19]
```diff
@@ -825,6 +825,41 @@ public Person findById(String id) {
 ----
 ====
 
+== Optimistic Lock Control
+
+Like we have in JPA/Hibernate, optimistic lock control is now supported, for this you just need to use the `@Version`, and panache does the rest.
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply docs/src/main/asciidoc/mongodb-panache.adoc:830]: done

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/exception/OptimisticLockException.java [hunk 20]
```diff
@@ -0,0 +1,12 @@
+package io.quarkus.mongodb.panache.common.exception;
+
+public class OptimisticLockException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/exception/OptimisticLockException.java:3]: sure, you are right, sorry.
Done

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 12 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:115]: adjusted.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 13 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:219]: @loicmathieu since we always return a `Uni<Void>`, we can ignore this missing delete.
I left this there for cases where you are trying to delete something that was recently updated by another user, it is just like an alert "hey, someone updated the info you want to delete, are you sure?".

This is there because I had a case where it was useful, but I can remove, what do you think?

The point @evanchooly said about trying to delete with a non existent id makes sense for me, and I think is better to remove this check when deleting.

By return something other than void, I think void is a standard in frameworks, panache/hibernante/jpa, all works in this way, I don't think we should change this

Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:219]: I've just remove this check here and for MongoOperations.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 11 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java:70]: I think we should raise an error, just adjusted here.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 14 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:244]: @evanchooly I adjusted the code to reset the version when the update fails, so the user will be able to recovery since there is no changes in the version.

@loicmathieu I really appreciate any tip here about the names/refactoring to make this easier to read, before I named this "of()" since it is a constructor, now it is "handle", but it is not cool for me.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java [hunk 21]
```diff
@@ -235,25 +241,40 @@ public Uni<Void> nullUni() {
...
+        }
         return collection.insertMany(entities).onItem().ignore().andContinueWithNull();
     }
 
     private Uni<Void> update(ReactiveMongoCollection collection, Object entity) {
+        return update(collection, entity, Optional.empty());
+    }
+
+    private Uni<Void> update(ReactiveMongoCollection collection, Object entity,
+            Optional<VersionHandler> versionHandlerOptional) {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:260]: makes sense, fixed.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java [hunk 22]
```diff
@@ -287,10 +312,15 @@ private Uni<Void> persistOrUpdate(ReactiveMongoCollection collection, List<Objec
...
-                BsonDocument query = new BsonDocument().append(ID, id);
-                bulk.add(new ReplaceOneModel(query, entity,
-                        new ReplaceOptions().upsert(true)));
+                //insert/update with user provided ID or update
+                VersionHandler versionHandler = VersionHandler.handle(entity, document);
+                if (versionHandler.containsVersionAnnotationAndValue()) {
+                    //it will ignore in silence not matching updates.
+                    bulk.add(new ReplaceOneModel(versionHandler.query, entity));
+                } else if (!versionHandler.containsVersionValue) {
+                    //only try upsert when there is no version value
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:321]: I've changed to handle this scenario, when we have versioned entities, we call the `update` method that check version and throw the OptimisticLockException, otherwise I call `collection.bulkWrite(bulk);`.
The drawback here is to no been able to call `bulkWrite` for all and get noticed of every single entity that could be failed in the update.
Do you have any suggestion to better handle this?

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 14 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java:244]: I had an idea here, I'm refactoring some things to make this clear :)

Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java]: @loicmathieu what I did was create the `EntityVersionInfo`, and auxiliar class that has info about version, and query to update. The `VersionHandler` produces an `EntityVersionInfo` and handle the version value, increasing or rollback the same in failure situations.

@evanchooly I've changed a lot this code, now every time you have the version annotation, it follows another path, that guarantee the state of the entity for all cases such as persist/persistOrUpdate/update, let me know your considerations.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java]: I've changed this, if we have version annotation, it follows another path and the state is guaranteed

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java [hunk 23]
```diff
@@ -235,25 +241,40 @@ public Uni<Void> nullUni() {
...
     private Uni<Void> persist(ReactiveMongoCollection collection, Object entity) {
+        VersionHandler.handle(entity, getBsonDocument(collection, entity));
         return collection.insertOne(entity).onItem().ignore().andContinueWithNull();
     }
 
     private Uni<Void> persist(ReactiveMongoCollection collection, List<Object> entities) {
+        for (Object entity : entities) {
+            VersionHandler.handle(entity, getBsonDocument(collection, entity));
+        }
         return collection.insertMany(entities).onItem().ignore().andContinueWithNull();
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java]: I've changed this, if we have version annotation, it follows another path and the state is guaranteed, let me know what do you think.

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Error on set the version value");
+        }
+    }
+
+    /**
+     * Utility method to check if the during an update there are documents not updated.
+     * Throws OptimisticLockException when found documents not updated.
+     */
+    public void checkHasNotAffectedResults(Long quantityResultAffected) {
```
[the remark this answers is not in the document]
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java]: Changed

File: extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java (hunk 11 above)
Comment (diogocarleto) [reply extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java]: @loicmathieu do you remember any place in panache making use of this? I can change here, btw I'd like to ask you for review again, and if all is good, can I change to jandex in the next PR?

**Comment by cescoffier** [comment]
Hum... something went terribly wrong with a rebase / update.

**Comment by diogocarleto** [comment]
I handled lots of files manually past week,would you like I close this and reopen the PR? Let me know what is the best approach for this :)

**Comment by cescoffier** [comment]
You can rebase from the main branch. Otherwise, yes, just recreate it.

**Comment by geoand** [comment]
Closing in favor of https://github.com/quarkusio/quarkus/pull/27966

**Comment by vsevel** [comment]
> Closing in favor of https://github.com/quarkusio/quarkus/pull/27966

hello @geoand . closing in favor in itself?
(we are very interested in this feature)

**Comment by geoand** [comment]
Thanks for catching the mistake!

I mean in favor of https://github.com/quarkusio/quarkus/pull/30144

**Comment by diogocarleto** [comment]
@geoand this PR is outdated, I can for sure review and update the PR if there is any interest in moving forward with this. 

In the past was told me that Panache was not designed to be an ODM ( what I totally disagree). I think the entire discussion is here in the comments.

BTW, we have an out of box implementation of this running in production for many years already in at least 2 projects, without any issue. For sure I’d love to see this inside Quarkus.

Let me know if I can work in this again. 

Best,

**Comment by geoand** [comment]
Here is my proposal: Start a discussion on the quarkus-dev mailing list mentioning the current state, what you've done, what the challenges are and what you proposal is so more people can chime in.

Thanks

**Comment by loicmathieu** [comment]
Yes, a discussion is a good starting point.
As far as I remember, the main point was that it has a high surface, this change modifies a lot of code so I was asking if something based on a CommandListener would be possible instead of updating MongoOperations.

**Comment by lyrst** [comment]
Started a conversation in the zulip-chat. https://quarkusio.zulipchat.com/#narrow/channel/187038-dev/topic/Quarkus.20.2B.20MongoDB.3A.20Support.20for.20optimistic.20locking.20in.20Panache/near/498089101

So maybe we can continue the discussion there.
